### ISSUE TITLE
Question on reconstruction loss implementation

### ISSUE THREAD

**Issue by DorisMai** [Hekstra-Lab/careless-examples#9]
In careless_zero.py, the log likelihood term in ELBO, $\log P(I|F^2\Sigma)$ is implemented as 
line 56 `likelihood = tfd.Normal(loc=intensities, scale=uncertainties)`
...
line 72 `log_likelihood = tf.reduce_sum(likelihood.log_prob(F * F * Sigma))`

I don't understand why this is implementing $\log P(I|F^2\Sigma)$ rather than $\log P(F^2\Sigma|I)$. I am new to VAE and ELBO, and any help is very appreciated.

**Comment by kmdalton** [comment]
Hi @DorisMai, thanks for a very insightful question. In this case we use an empirically estimated uncertainty, $\sigma_I$, with the log likelihood function, $\log \mathrm{Normal}$. We're interested in computing the expected log likelihood of the data given the model, $\mathbb{E}_{q}  \log \mathrm{Normal}(I | F^2\Sigma, \sigma_I)$. We do so by taking Monte Carlo samples from the variational distributions for $F$ and $\Sigma$ which I collectively refer to as $q$. 

As you note, the code actually implements a different likelihood $\mathbb{E}_{q}  \log P(F^2\Sigma | I, \sigma_I)$. However, this is okay, because $\log \mathrm{Normal} (x | \mu, \sigma)$ is invariant under permutation of $x, \mu$. In our case, for a random sample $z \sim q$,
```math
\log P(I | z, \sigma_I) = \log P(z | I, \sigma_I)
```
That is to say, one could compute either likelihood because they are equivalent. All of the likelihoods in `careless` are symmetric in this way. However, this is not generally the case. It would be reasonable to change the likelihood to be asymmetric; let's say Rician or LogNormal for the sake of argument. In which case, you would want to change the order of the calculation in order to reflect the likelihood you desired. 

In the context of `careless_zero.py`, this could be achieved by constructing the likelihood distribution in the `minus_elbo` function — as I am sure you have already surmised. Nothing stops one from doing the following,
```python
#Evaluate the elbo
def minus_elbo():
    z = q.sample(mc_samples)
    F = tf.gather(z, miller_id, axis=1)
    loc, scale = tf.unstack(NN(metadata), axis=1)
    Sigma = tfd.Normal(loc, scale).sample(mc_samples)
    likelihood = tfd.Normal(loc=F*F*Sigma, scale=uncertainties)
    log_likelihood = tf.reduce_sum(likelihood.log_prob(intensities))
    log_p_z = tf.where(centric, p_centric.log_prob(z), p_acentric.log_prob(z))
    log_q_z = q.log_prob(z)
    kl_div = tf.reduce_sum(log_q_z - log_p_z)
    return -log_likelihood + kl_div
```


So, really there is nothing deep here. Just an implementation choice I made. Perhaps it was not the wisest choice from a pedagogical standpoint. Perhaps I will change it to look more like the code block above.

**Comment by DorisMai** [comment]
Thank you for the clarification and the code example! This is very helpful.
